lib/player.js

```diff
@@ -2001,6 +2016,21 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
     });
     this.abrManager_.setMediaElement(mediaElement);
 
+    // Pick the initial streams to play.
+    // Unless the user has already picked a variant, anyway, by calling
+    // selectVariantTrack before this loading stage.
+    let initialVariant = null;
+    const activeVariantTrack = this.getVariantTracks().find((t) => t.active);
+    if (!activeVariantTrack) {
+      initialVariant = this.chooseVariant_();
```

Comment:
I'm a little concerned about doing this before the app's codec preferences are applied.  I think choice of codec should still come before choice of initial variant.

I see a comment below that duration is needed earlier, and I presume this is needed by playhead.  But as far as I can tell (I could be wrong), playhead isn't needed by StreamingEngine directly.  StreamingEngine appears to need duration only after start() is called, but no sooner.

Could you put all this back the way it was, and move playhead's creation until right before switchVariant_(), and load the variant right before creating playhead and starting StreamingEngine?  That would mean you don't have to load anything until both after initial the variant is chosen and after the streaming event is handled, giving the app a chance to override it.  Would that work?

Reply:
Huh. Ended up being simpler then I expected.